### PR TITLE
Noc blocks memory map initial implementation

### LINKED ISSUE #709: [Bug Report] Command Line `brxy` always shows `(L1)`
**Describe the bug**
I was recently using the CLI for tt-exalens, and was getting confused because the printout always shows the data as resident in L1, even if it isn't (see reproduction steps for example outputs).

**To Reproduce**
`brxy 0-0 0x1167DB60 32`
```
0-0 (L1) : 0x1167db60 (128 bytes)
<data at address starts here>
```
The above is confusing because this is actually in DRAM, not L1

`brxy 0,0 0x1167DB60 32`
```
0,0 (L1) : 0x1167db60 (128 bytes)
<data at address starts here>
```
I believe the above is in L1, but may be outside the bounds of L1, so I'm not entirely sure what is actually being returned

`brxy e0,0 0x1167DB60 32`
```
0,0 (L1) : 0x1167db60 (128 bytes)
<data at address starts here>
```
Similarly confusing, the above should be in ETH, not L1

**Expected behavior**
The proper location should be shown instead.

### PR DESCRIPTION
**Pull request by onenezicTT**
The idea behind this PR is to start some kind of a map that maps tiles' memory regions accessible via NoC which brxy will be able to use easily, in an ideal world this map will be extended with additional info so that it can be generally used as a source for private and raw addresses as well.
This PR also fixes a bug reported where brxy always shows L1 as the source of data when reading anything other than dram channels. Not all memory regions are known and will be added as follow ups. 

The output format is:
`"<core> : 0x<addr> (<total_bytes> total bytes)`
Followed by one or more blocks as `(<BLOCK_NAME>) : 0x<addr> (<bytes>)` with hex dump

Closes #709 

Follow-ups:
More tile types (arc/eth?)
More regions (example where registers are mapped etc.)

---

> [!NOTE]
> Add a MemoryMap for NoC-accessible regions, integrate it into NOC blocks, and enhance brxy burst-read to label dumps by memory block; include unit tests.
> 
> - **Core**:
>   - Add `MemoryMap` with name/address lookups and sorted NoC regions (`ttexalens/memory_map.py`).
>   - Expose optional `memory_map` via `NocBlock.get_memory_map()`.
> - **CLI**:
>   - Update `burst_read_xy` to use `MemoryMap` for burst reads: prints overall range and splits output into labeled blocks; sampling mode includes block names. Add `print_memory_block` helper.
> - **Hardware**:
>   - Populate `memory_map` in blocks with known regions:
>     - Blackhole: `functional_worker_block` (L1, BRISC/NCRISC/TRISC data privates), `dram_block` (L1, DRAM bank).
>     - Wormhole: `functional_worker_block` (L1), `dram_block` (DRAM bank).
>     - Quasar: `functional_worker_block` (L1, neo[0-3] TRISC data privates).
> - **Tests**:
>   - Add unit tests for `MemoryMap` covering init, sorting, lookups, boundaries, and edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0af952662fadc59d603dec138a401cd64f7e1fa3. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: ttexalens/noc_memory_map.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
+
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+from typing import TypedDict
+
+
+class MemoryRegion(TypedDict):
```
Comment (tt-vjovanovic) [inline ttexalens/noc_memory_map.py]: Consider using `MemoryBlock`

File: ttexalens/hardware/blackhole/dram_block.py [hunk 2]
```diff
@@ -185,6 +186,14 @@ def __init__(self, location: OnChipCoordinate):
         self.register_store_noc0 = RegisterStore(register_store_noc0_initialization, self.location)
         self.register_store_noc1 = RegisterStore(register_store_noc1_initialization, self.location)
 
+        # Create NOC memory map
+        self.noc_memory_map = NocMemoryMap(
+            {
+                "l1": {"noc_address": 0x00000000, "size": 4 * 1024},
```
Comment (tt-vjovanovic) [inline ttexalens/hardware/blackhole/dram_block.py]: This is repeated information. We should use `MemoryBlock` from `l1` variable.

File: ttexalens/hardware/blackhole/functional_worker_block.py [hunk 3]
```diff
@@ -178,6 +179,18 @@ def __init__(self, location: OnChipCoordinate):
         self.register_store_noc0 = RegisterStore(register_store_noc0_initialization, self.location)
         self.register_store_noc1 = RegisterStore(register_store_noc1_initialization, self.location)
 
+        # Create NOC memory map
+        self.noc_memory_map = NocMemoryMap(
+            {
+                "l1": {"noc_address": 0x00000000, "size": 1536 * 1024},
+                "brisc_private_memory": {"noc_address": 0xFFB14000, "size": 8 * 1024},
```
Comment (tt-vjovanovic) [inline ttexalens/hardware/blackhole/functional_worker_block.py]: `brisc_data_private_memory`

File: ttexalens/cli_commands/burst_read_xy.py [hunk 4]
```diff
@@ -110,26 +110,92 @@ def print_a_burst_read(
...
-        print(f"{da._id}\n{util.dump_memory(addr, da.data, bytes_per_entry, 16, is_hex)}")
+
+        # Print overall header
+        print(f"{core_loc_str} : 0x{addr:08x} ({word_count * 4} total bytes)")
+
+        if memory_map is not None:
+            i = 0
+            while i < word_count:
+                word_addr = addr + i * 4
+                region_name = memory_map.get_region_by_noc_address(word_addr)
```
Comment (tt-vjovanovic) [inline ttexalens/cli_commands/burst_read_xy.py]: `get_region_by_address`

File: ttexalens/noc_memory_map.py (hunk 1 above)
Comment (onenezicTT) [reply ttexalens/noc_memory_map.py]: I was originally using that though somehow moved away from it, but it makes sense to use it

File: ttexalens/noc_memory_map.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    """A simple memory region descriptor."""
+
+    noc_address: int
+    size: int
+
+
+class NocMemoryMap:
```
[the remark this answers is not in the document]
Comment (onenezicTT) [reply ttexalens/noc_memory_map.py]: Let's revert to that

File: ttexalens/cli_commands/burst_read_xy.py (hunk 4 above)
Comment (onenezicTT) [reply ttexalens/cli_commands/burst_read_xy.py]: Even though currently we do searches by noc_address?

**Comment by onenezicTT** [comment]
@tt-vjovanovic please let me know if some of the follow-ups listed in the PR description make sense to do in this PR :)

File: ttexalens/hardware/quasar/functional_worker_block.py [hunk 6]
```diff
@@ -49,6 +50,50 @@ def __init__(self, location: OnChipCoordinate):
             risc_base_start_address=0x00030000,
         )
 
+        # Quasar functional workers always have data private memory for all trisc cores
+        assert self.neo0.trisc0.data_private_memory is not None
```
Comment (tt-vjovanovic) [inline ttexalens/hardware/quasar/functional_worker_block.py]: These asserts look horrible. Can we tag somehow line that this is always `not None`?

File: ttexalens/memory_map.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        """
+        Initialize memory map with named blocks.
+
+        Args:
+            blocks: Dictionary mapping block names to their descriptors
+        """
+
+        self.blocks: dict[str, MemoryBlock] = blocks
+        # Note: sortedBlocks only contains blocks with noc_address
+        self.sortedBlocks: list[tuple[MemoryBlock, str]] = sorted(
```
Comment (tt-vjovanovic) [inline ttexalens/memory_map.py]: `sorted_blocks_by_noc_address`

File: ttexalens/cli_commands/burst_read_xy.py [hunk 8]
```diff
@@ -105,31 +106,101 @@ def print_a_read(core_loc_str, addr, val, comment=""):
...
+                block_name = "?"
+
+            block_header = f"{core_loc_str} ({block_name})"
+
             values = {}
             print(f"Sampling for {sample / word_count} second{'s' if sample != 1 else ''}...")
             t_end = time.time() + sample / word_count
             while time.time() < t_end:
-                val = read_word_from_device(core_loc, addr + 4 * i, device_id, context=context)
+                val = read_word_from_device(core_loc, word_addr, device_id, context=context)
```
Comment (adjordjevic-TT) [inline ttexalens/cli_commands/burst_read_xy.py]: Can you change every reference of `core_loc` to `location`. We switched a while back because we no longer call NOC nodes cores.

File: ttexalens/cli_commands/burst_read_xy.py [hunk 9]
```diff
@@ -105,31 +106,101 @@ def print_a_read(core_loc_str, addr, val, comment=""):
...
+                if memory_block_name is not None:
+                    # Get block info and calculate how many words fit in this known region
+                    memory_block = memory_map.get_block_by_name(memory_block_name)
+                    assert memory_block is not None, f"Memory block '{memory_block_name}' not found in map but expected"
+                    assert (
+                        memory_block.address.noc_address is not None
+                    ), f"Memory block '{memory_block_name}' has no NOC address"
+                    memory_block_start = memory_block.address.noc_address
+                    memory_block_end = memory_block_start + memory_block.size
+                    remaining_words_in_block = (memory_block_end - word_addr) // 4
```
[the remark this answers is not in the document]
Comment (tt-vjovanovic) [reply ttexalens/cli_commands/burst_read_xy.py]:
```
remaining_words_in_block = min((memory_block_end - word_addr) // 4, 1)
```

**Review by adjordjevic-TT — APPROVED** [review]

File: ttexalens/hardware/noc_block.py [hunk 10]
```diff
@@ -7,19 +7,23 @@
...
     from ttexalens.register_store import RegisterStore
+    from ttexalens.memory_map import MemoryMap
 
 
 class NocBlock:
     def __init__(self, location: OnChipCoordinate, block_type: str, debug_bus: DebugBusSignalStore | None = None):
         self.location = location
         self.block_type = block_type
         self.debug_bus = debug_bus
+        self.memory_map: MemoryMap | None = None
```
[the remark this answers is not in the document]
Comment (onenezicTT) [reply ttexalens/hardware/noc_block.py]: Unfortunatelly cannot get passed testing for None without something, and that something is those tags...

**Review by tt-vjovanovic — APPROVED** [review]

File: ttexalens/memory_map.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    def map_blocks(self, blocks: list[tuple[str, MemoryBlock]]) -> None:
+        for name, memory_block in blocks:
+            self.map_block(name, memory_block)
+
+    def get_block_name_by_noc_address(self, noc_address: int) -> str | None:
+        interval: Interval = self.noc_address_to_block_name_mapping.at(noc_address)
+        if interval is None:
+            return None
+
+        # Assuming no overlapping blocks, return the first found
```
[the remark this answers is not in the document]
Comment (onenezicTT) [reply ttexalens/memory_map.py]: Good idea, we should be stricter here

File: ttexalens/memory_map.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+class MemoryMap:
+    """Catalog of memory blocks with address or name based lookup."""
+
+    def __init__(self):
+        self.noc_address_to_block_name_mapping = IntervalTree()
+        self.name_to_block_mapping: dict[str, MemoryBlock] = {}
+
+    def map_block(self, name: str, memory_block: MemoryBlock) -> None:
+        # Currently doing only NoC address mapping
+        assert memory_block.address.noc_address is not None, "Memory block must have NoC address to be mapped"
```
[the remark this answers is not in the document]
Comment (onenezicTT) [reply ttexalens/memory_map.py]: It's probably fine for us to not be strict here on asserts and just not add it

File: ttexalens/cli_commands/burst_read_xy.py [hunk 13]
```diff
@@ -101,35 +102,93 @@ def process_device(device_id):
...
+    da = DataArray(f"{header} : 0x{start_addr:08x} ({len(data) * 4} bytes)", 4)
+    da.data = data
+    if bytes_per_entry != 4:
+        da.to_bytes_per_entry(bytes_per_entry)
+    print(f"{da._id}\n{util.dump_memory(start_addr, da.data, bytes_per_entry, 16, is_hex)}")
 
 
 def print_a_burst_read(
-    device_id, core_loc, addr, core_loc_str, word_count=1, sample=1, print_format="hex32", context=None
+    device_id, location, addr, location_str, word_count=1, sample=1, print_format="hex32", context=None
```
[the remark this answers is not in the document]
Comment (onenezicTT) [reply ttexalens/cli_commands/burst_read_xy.py]: Didn't know that, cool
